Replace role system (RBAC) with permissions-based DB roles

AlanCoding opened the pull request:
##### SUMMARY
This replaces https://github.com/ansible/awx/pull/14735 as we have introduced a feature branch for this purpose.

This makes use of https://github.com/ansible/django-ansible-base/pull/45 and the intent is to have that merged before this.

New endpoints are exposed from DAB. These are
 - role definitions
 - role user assignments
 - role team assignments

The routers/views/serializers for these new endpoints live in DAB, not in this patch. This is mostly concerned with the "translation" layer to make the new roles look kind of like the old roles. Generally this is intended to be a backward-compatible change.

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
 - API


##### ADDITIONAL INFORMATION

AlanCoding committed:
Replace role system with permissions-based DB roles

Develop ability to list permissions for existing roles

Create a model registry for RBAC-tracked models

Write the data migration logic for creating
  the preloaded role definitions

Write migration to migrate old Role into ObjectRole model

This loops over the old Role model, knowing it is unique
  on object and role_field

Most of the logic is concerned with identifying the
  needed permissions, and then corresponding role definition

As needed, object roles are created and users then teams
  are assigned

Write re-computation of cache logic for teams
  and then for object role permissions

Migrate new RBAC internals to ansible_base

Migrate tests to ansible_base

Implement solution for visible_roles

Expose URLs for DAB RBAC

AlanCoding committed:
Cast ObjectRole object_id to int, very wrong, tmp fix

AlanCoding committed:
Cache organization child evaluations and remove hacks

AlanCoding committed:
Bump number of allowed endpoints (#14956)

AlanCoding committed:
[DAB RBAC] Re-implement system auditor as a singleton role in new system (#14963)

* Add new enablement settings from DAB RBAC

* Initial implementation of system auditor as role without testing

* Fix system auditor role, remove duplicate assignments

* Make the system auditor role managed

* Flake8 fix

* Remove another thing from old solution

* Fix a few test failures

* Add extra setting to disable custom system roles via API

* Add test for custom role prohibition

fosterseth commented on awx/api/generics.py:
```diff
@@ -472,7 +476,11 @@ def skip_related_name(name):
+        super().perform_create(serializer)
+        if serializer.Meta.model in permission_registry.all_registered_models:
+            if self.request and self.request.user:
+                give_creator_permissions(self.request.user, serializer.instance)
```
trying to think of other resources that are created in non-conventional ways

ReceptorAddress in the create_or_update method of the InstanceSerializer
- I think this is admin-only anyways so maybe doesn't matter too much

BulkHost and BulkLaunch endpoints

AlanCoding replied:
Good question. Those bulk endpoints create hosts & jobs. Hosts defer to inventory permission, and jobs defer to job template (or organization as fallback) permission. So no permission action is needed in those cases.

AlanCoding committed:
Use AWX base view to make unauth requests 401 (#14981)

AlanCoding committed:
Minor RBAC test fix (#14982)

AlanCoding committed:
Adopt internal DAB RBAC Permission model (#14994)

AlanCoding committed:
Bump migration number for RBAC branch

AlanCoding committed:
[RBAC] Fix migration for created and modified field changes (#14999)

Fix migration for created and modified field changes

AlanCoding committed:
[RBAC] Fix server error from delete capability of approvals (#15002)

Fix server error from delete capability of approvals

AlanCoding committed:
Generalize can_delete solution, use devel DAB (#15009)

* Generalize can_delete solution, use devel DAB

* Fix bug where model was used instead of model_name

* Linter fixes

jbradberry commented on awx/api/views/__init__.py:
```diff
@@ -536,10 +540,12 @@ class InstanceGroupAccessList(ResourceAccessList):
 
 
 class InstanceGroupObjectRolesList(SubListAPIView):
+    deprecated = True
```
Are we really going to deprecate everything?  That seems pretty hostile to community users.

(also, this one seems to have `deprecated = True` twice.)

PabloHiro replied:
Talking about deprecation, if I understood you correctly `access_list` should be marked as deprecated as well right?

AlanCoding replied:
Put up https://github.com/ansible/awx/pull/15068, which should address the things mentioned in both of the above 2 comments.

And yes, these will be deprecated. Obviously we won't just go break the UI. We will push the new UI to the community somehow and stop maintaining the old UI. Some time after that, yes, we will remove these endpoints, and we will delete the old `Role` tables that are used for the backward compatibility. Everything you could do with these you can do with the new endpoints.

AlanCoding committed:
[RBAC] Fix known issues with backward compatible access_list (#15052)

* Remove duplicate access_list entries for direct team access

* Revert test changes for superuser in access_list

fosterseth committed:
AWX Collections for DAB RBAC

Adds new modules for CRUD operations on the
following endpoints:

- api/v2/role_definitions
- api/v2/role_user_assignments
- api/v2/role_team_assignments

Note: assignment is Create or Delete only

Additional changes:
- Currently DAB endpoints do not have "type"
field on the resource list items. So this modifies
the create_or_update_if_needed to allow manually
specifying item type.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

AlanCoding committed:
Fix missing role membership when giving creator permissions (#15058)

AlanCoding committed:
[RBAC] Tweaks to reflect what endpoints are deprecated (#15068)

Tweaks to reflect what endpoints are deprecated

AlanCoding committed:
[RBAC] Fix bug where team could not be given read_role to other team (#15067)

* Fix bug where team could not be given read_role to other team

* Avoid unwanted triggers of parentage granting

* Restructure signal structure

* Fix another bug unmasked by team member permission fix

* Changes to live with test writing

* Use equality as opposed to string "in"

from Seth in review comment

Co-authored-by: Seth Foster <fosterseth@users.noreply.github.com>

---------

Co-authored-by: Seth Foster <fosterseth@users.noreply.github.com>

AlanCoding committed:
[RBAC] Rename managed role definitions, and move migration logic here (#15087)

* Rename managed role definitions, and move migration logic here

* Fix naming capitalization

AlanCoding committed:
Make custom urls work with RBAC

AlanCoding committed:
[RBAC] Update related name to reflect upstream DAB change (#15093)

Update related name to reflect upstream DAB change
